src/auth.ts

```diff
@@ -0,0 +1,449 @@
+import * as opn from "open";
+import * as path from "path";
+import * as portfinder from "portfinder";
+import * as url from "url";
+import * as util from "util";
+
+import * as api from "./api";
+import { configstore } from "./configstore";
+import { FirebaseError } from "./error";
+import * as logger from "./logger";
+import { prompt } from "./prompt";
+import * as scopes from "./scopes";
+
+/* eslint-disable camelcase */
+// The wire protocol for an access token returned by Google.
+// When we actually refresh from the server we should always have
+// these optional fields, but when a user passes --token we may
+// only have access_token.
+interface Tokens {
+  id_token?: string;
+  access_token: string;
+  refresh_token?: string;
+  scopes?: string[];
+}
+
+interface TokensWithExpiration extends Tokens {
+  expires_at?: number;
+}
+
+interface TokensWithTTL extends Tokens {
+  expires_in?: number;
+}
+
+interface UserCredentials {
+  user: string | { [key: string]: any };
+  tokens: TokensWithExpiration;
+  scopes: string[];
+}
+
+// https://docs.github.com/en/developers/apps/authorizing-oauth-apps
+interface GitHubAuthResponse {
+  access_token: string;
+  scope: string;
+  token_type: string;
+}
+/* eslint-enable camelcase */
+
+// Typescript emulates modules, which have constant exports. We can
+// overcome this by casting to any
+// Should be fixable after https://github.com/http-party/node-portfinder/pull/115
+(portfinder as any).basePort = 9005;
+
+function open(url: string) {
+  opn(url).catch((err) => {
+    logger.debug("Unable to open URL: " + err.stack);
+  });
+}
+
+// Always create a new error so that the stack is useful
+function invalidCredentialError(): FirebaseError {
+  return new FirebaseError(
+    "Authentication Error: Your credentials are no longer valid. Please run " +
+      clc.bold("firebase login --reauth") +
+      "\n\n" +
+      "For CI servers and headless environments, generate a new token with " +
+      clc.bold("firebase login:ci"),
+    { exit: 1 }
+  );
+}
+
+const FIFTEEN_MINUTES_IN_MS = 15 * 60 * 1000;
+const SCOPES = [
+  scopes.EMAIL,
+  scopes.OPENID,
+  scopes.CLOUD_PROJECTS_READONLY,
+  scopes.FIREBASE_PLATFORM,
+  scopes.CLOUD_PLATFORM,
+];
+
+const _nonce = Math.floor(Math.random() * (2 << 29) + 1).toString();
+const getPort = portfinder.getPortPromise;
+
+// in-memory cache, so we have it for successive calls
+let lastAccessToken: TokensWithExpiration | undefined;
+
+function getCallbackUrl(port?: number) {
+  if (typeof port === "undefined") {
+    return "urn:ietf:wg:oauth:2.0:oob";
+  }
+  return "http://localhost:" + port;
+}
+
+function queryParamString(args: Object) {
+  const tokens: string[] = [];
+  for (const [key, value] of Object.entries(args)) {
+    if (typeof value === "string") {
+      tokens.push(key + "=" + encodeURIComponent(value));
+    }
+  }
+  return tokens.join("&");
+}
+
+function getLoginUrl(callbackUrl: string, userHint?: string | null) {
```

Comment:
IIRC this is because there was an explicit null back when we had unified login for Google and GitHub (the latter passed an explicit `null`. Will try removing.